packages/messageport/tizen/src/messageport.cc

```diff
@@ -5,234 +5,161 @@
+    ret = message_port_check_trusted_remote_port(app_id_.c_str(), name_.c_str(),
+                                                 &exist);
   } else {
-    ret = message_port_check_remote_port(remote_app_id.c_str(),
-                                         port_name.c_str(), port_check);
+    ret =
+        message_port_check_remote_port(app_id_.c_str(), name_.c_str(), &exist);
   }
 
-  LOG_DEBUG("message_port_check_%s_remote_port (%s): %s",
-            is_trusted ? "trusted" : "", port_name.c_str(),
-            *port_check ? "true" : "false");
-
-  return CreateResult(ret);
+  if (ret != MESSAGE_PORT_ERROR_NONE) {
+    return MessagePortError(ret);
+  }
+  return exist;
 }
 
-MessagePortResult MessagePortManager::Send(std::string& remote_app_id,
-                                           std::string& port_name,
-                                           flutter::EncodableValue& message,
-                                           bool is_trusted) {
-  LOG_DEBUG("Send (%s, %s), trusted: %s", remote_app_id.c_str(),
-            port_name.c_str(), is_trusted ? "yes" : "no");
-  bundle* bundle = nullptr;
-  MessagePortResult result = PrepareBundle(message, bundle);
-  if (!result) {
-    return result;
+std::optional<MessagePortError> RemotePort::Send(
+    std::unique_ptr<std::vector<uint8_t>> encoded_message) {
+  ErrorOr<bundle*> maybe_bundle = PrepareBundle(encoded_message.get());
+  if (maybe_bundle.has_error()) {
+    return maybe_bundle.error();
   }
+  bundle* bundle = maybe_bundle.value();
 
-  int ret;
-  if (is_trusted) {
-    ret = message_port_send_trusted_message(remote_app_id.c_str(),
-                                            port_name.c_str(), bundle);
+  int ret = MESSAGE_PORT_ERROR_NONE;
+  if (is_trusted_) {
+    ret = message_port_send_trusted_message(app_id_.c_str(), name_.c_str(),
+                                            bundle);
   } else {
-    ret = message_port_send_message(remote_app_id.c_str(), port_name.c_str(),
-                                    bundle);
+    ret = message_port_send_message(app_id_.c_str(), name_.c_str(), bundle);
   }
   bundle_free(bundle);
 
-  return CreateResult(ret);
+  if (ret != MESSAGE_PORT_ERROR_NONE) {
+    return MessagePortError(ret);
+  }
+
+  return std::nullopt;
 }
 
-MessagePortResult MessagePortManager::Send(std::string& remote_app_id,
-                                           std::string& port_name,
-                                           flutter::EncodableValue& message,
-                                           bool is_trusted, int local_port) {
-  LOG_DEBUG("Send (%s, %s), port: %d, trusted: %s", remote_app_id.c_str(),
-            port_name.c_str(), local_port, is_trusted ? "yes" : "no");
-  bundle* bundle = nullptr;
-  MessagePortResult result = PrepareBundle(message, bundle);
-  if (!result) {
-    return result;
+std::optional<MessagePortError> RemotePort::Send(
+    std::unique_ptr<std::vector<uint8_t>> encoded_message,
+    LocalPort* local_port) {
```

Comment:
> Also encoded_message shouldn't necessarily be a unique_ptr. It could be just const std::vector<uint8_t> & if you want to pass data without copying.

@swift-kim 
I think it is natural because the result of decoding with `StandardMessageCodec` is `unique_ptr`. or using raw pointer from `unique_ptr` is better for this case. also, it need to remove the `const` again to change `vector` to a `bundle`. how about this?
 
``` c++ 
std::unique_ptr<std::vector<uint8_t>> encoded_message =
        flutter::StandardMessageCodec::GetInstance().EncodeMessage(message);
// ...
remote_port.SendWithLocalPort(encoded_message.get()) // remote_port.SendWithLocalPort(*(encoded_message.get()))

```

Reply:
This totally makes sense:
```patch
--- a/packages/messageport/tizen/src/messageport.cc
+++ b/packages/messageport/tizen/src/messageport.cc
@@ -121,9 +121,8 @@ std::optional<MessagePortError> RemotePort::Send(
 }

 std::optional<MessagePortError> RemotePort::Send(
-    std::unique_ptr<std::vector<uint8_t>> encoded_message,
-    LocalPort* local_port) {
-  ErrorOr<bundle*> maybe_bundle = PrepareBundle(encoded_message.get());
+    const std::vector<uint8_t>& encoded_message, LocalPort* local_port) {
+  ErrorOr<bundle*> maybe_bundle = PrepareBundle(encoded_message);
   if (maybe_bundle.has_error()) {
     return maybe_bundle.error();
   }
@@ -147,14 +146,14 @@ std::optional<MessagePortError> RemotePort::Send(
 }

 ErrorOr<bundle*> RemotePort::PrepareBundle(
-    std::vector<uint8_t>* encoded_message) {
+    const std::vector<uint8_t>& encoded_message) {
   bundle* bundle = bundle_create();
   if (!bundle) {
     return MessagePortError(MESSAGE_PORT_ERROR_OUT_OF_MEMORY);
   }

-  int ret = bundle_add_byte(bundle, "bytes", encoded_message->data(),
-                            encoded_message->size());
+  int ret = bundle_add_byte(bundle, "bytes", encoded_message.data(),
+                            encoded_message.size());

   if (ret != BUNDLE_ERROR_NONE) {
     bundle_free(bundle);

--- a/packages/messageport/tizen/src/messageport_tizen_plugin.cc
+++ b/packages/messageport/tizen/src/messageport_tizen_plugin.cc
@@ -249,9 +249,9 @@ class MessageportTizenPlugin : public flutter::Plugin {
         result->Error("Invalid arguments");
         return;
       }
-      error = remote_port.Send(std::move(encoded_message), local_port);
+      error = remote_port.Send(*encoded_message, local_port);
     } else {
-      error = remote_port.Send(std::move(encoded_message));
+      error = remote_port.Send(*encoded_message);
     }

     if (error.has_value()) {
```

The methods are synchronous and read-only (they do not modify the data or store it elsewhere), so transferring ownership is not very necessary in this case.